File: app/src/main/java/dev/adamsjack/beanshelf/AppViewModel.kt
package dev.adamsjack.beanshelf

import android.app.Application
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import dev.adamsjack.beanshelf.data.BeanPack
import dev.adamsjack.beanshelf.data.BeanStore
import dev.adamsjack.beanshelf.data.PhotoStore
import dev.adamsjack.beanshelf.model.Bean
import dev.adamsjack.beanshelf.model.Brew
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

/** In-app navigation: three screens, no nav library. */
sealed interface Screen {
    data object Shelf : Screen
    data object Leaderboard : Screen
    data object Social : Screen
    data class Detail(val beanId: String) : Screen
    /** beanId == null → new bag. Back returns to Detail when editing, Shelf when adding. */
    data class Edit(val beanId: String?) : Screen
}

class AppViewModel(app: Application) : AndroidViewModel(app) {

    private val store = BeanStore(app)

    private val _beans = MutableStateFlow<List<Bean>>(emptyList())
    val beans = _beans.asStateFlow()

    var nav by mutableStateOf<Screen>(Screen.Shelf)

    init {
        viewModelScope.launch { _beans.value = store.load() }
    }

    fun bean(id: String): Bean? = _beans.value.firstOrNull { it.id == id }

    fun upsertBean(bean: Bean) {
        val old = bean(bean.id)
        if (old != null && old.photoPath != null && old.photoPath != bean.photoPath) {
            PhotoStore.deletePhoto(old.photoPath)
        }
        if (old != null && old.backPhotoPath != null && old.backPhotoPath != bean.backPhotoPath) {
            PhotoStore.deletePhoto(old.backPhotoPath)
        }
        _beans.value = (_beans.value.filterNot { it.id == bean.id } + bean)
            .sortedByDescending { it.createdAt }
        persist()
    }

    fun deleteBean(id: String) {
        bean(id)?.let {
            PhotoStore.deletePhoto(it.photoPath)
            PhotoStore.deletePhoto(it.backPhotoPath)
        }
        _beans.value = _beans.value.filterNot { it.id == id }
        persist()
    }

    /** Imports a friend's .beanshelf file onto the shelf. */
    fun importBean(uri: android.net.Uri) {
        viewModelScope.launch {
            val bean = BeanPack.import(getApplication(), uri)
            if (bean != null) {
                _beans.value = (listOf(bean) + _beans.value).sortedByDescending { it.createdAt }
                persist()
                android.widget.Toast.makeText(
                    getApplication(),
                    "\"${bean.name.ifBlank { "Bean" }}\" added to your shelf",
                    android.widget.Toast.LENGTH_SHORT,
                ).show()
            } else {
                android.widget.Toast.makeText(
                    getApplication(), "That file isn't a bean pack", android.widget.Toast.LENGTH_SHORT,
                ).show()
            }
        }
    }

    fun addBrew(beanId: String, brew: Brew) {
        _beans.value = _beans.value.map {
            if (it.id == beanId) it.copy(brews = listOf(brew) + it.brews) else it
        }
        persist()
    }

    fun goBack() {
        nav = when (val s = nav) {
            is Screen.Edit -> if (s.beanId != null) Screen.Detail(s.beanId) else Screen.Shelf
            else -> Screen.Shelf
        }
    }

    private fun persist() {
        val snapshot = _beans.value
        viewModelScope.launch { store.save(snapshot) }
    }
}

File: ios/Beanshelf/AppViewModel.swift
import Foundation
import SwiftUI

/// Full-screen screens that cover the bottom bar (pushed over a tab).
enum Overlay: Equatable {
    case leaderboard
    case detail(String)
    /// beanId == nil → new bag. Back returns to Detail when editing, else the tab.
    case edit(String?)
}

@MainActor
final class AppViewModel: ObservableObject {

    @Published private(set) var beans: [Bean] = []
    /// Transient banner text ("added to your shelf" after an import).
    @Published var banner: String?

    @Published var tab: AppTab = .shelf
    @Published var overlay: Overlay?
    /// Set by a beanshelf://u/<username> deep link; consumed by the Feed tab.
    @Published var pendingProfile: String?

    private let store = BeanStore()

    func openProfile(_ username: String) {
        pendingProfile = username
        overlay = nil
        tab = .feed
    }

    func openDetail(_ id: String) { overlay = .detail(id) }
    func openEdit(_ id: String?) { overlay = .edit(id) }
    func openLeaderboard() { overlay = .leaderboard }
    func closeOverlay() { overlay = nil }

    func goBack() {
        switch overlay {
        case .edit(let beanId):
            overlay = beanId.map { .detail($0) }
        case .some:
            overlay = nil // Detail or Leaderboard → back to the tab
        case nil: // on a tab root — fall back to the Shelf tab
            pendingProfile = nil
            tab = .shelf
        }
    }

    func notify(_ text: String) {
        showBanner(text)
    }

    /// Every brew across the shelf, newest first — used to prefill equipment.
    var allBrews: [Brew] {
        beans.flatMap(\.brews).sorted { $0.timestamp > $1.timestamp }
    }

    func load() async {
        beans = await store.load()
    }

    func bean(_ id: String) -> Bean? {
        beans.first { $0.id == id }
    }

    func upsert(_ bean: Bean) {
        let old = self.bean(bean.id)
        if let old, let oldPhoto = old.photoFile, oldPhoto != bean.photoFile {
            PhotoStore.delete(oldPhoto)
        }
        if let old, let oldBack = old.backPhotoFile, oldBack != bean.backPhotoFile {
            PhotoStore.delete(oldBack)
        }
        beans = (beans.filter { $0.id != bean.id } + [bean])
            .sorted { $0.createdAt > $1.createdAt }
        persist()
    }

    func delete(id: String) {
        if let bean = bean(id) {
            PhotoStore.delete(bean.photoFile)
            PhotoStore.delete(bean.backPhotoFile)
        }
        beans.removeAll { $0.id == id }
        persist()
    }

    /// Imports a friend's .beanshelf file onto the shelf.
    func importBean(url: URL) async {
        if let bean = await BeanPack.import(url: url) {
            beans = ([bean] + beans).sorted { $0.createdAt > $1.createdAt }
            persist()
            showBanner("\"\(bean.name.isEmpty ? "Bean" : bean.name)\" added to your shelf")
        } else {
            showBanner("That file isn't a bean pack")
        }
    }

    func addBrew(beanId: String, brew: Brew) {
        beans = beans.map {
            var bean = $0
            if bean.id == beanId { bean.brews.insert(brew, at: 0) }
            return bean
        }
        persist()
    }

    private func showBanner(_ text: String) {
        banner = text
        Task {
            try? await Task.sleep(for: .seconds(2.5))
            if banner == text { banner = nil }
        }
    }

    private func persist() {
        let snapshot = beans
        Task { await store.save(snapshot) }
    }
}
